fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,118 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('api/todos')
+      .then(response => response.json())
+      .then(data => {
+        this.setState({
+          todos: data
+        });
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    */
+    fetch('api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => {
+        this.setState(state => {
+          state.todos.push(data);
+        });
```

Comment:
This "works" but it's not the intended way of using `setState`. If you pass a function to `setState`, it needs to return an object containing the property of state that you want to replace.
1. make a copy of the original array (see Array.prototyp.slice)
2. push the data (the new todo from the server) into the copy
3. `return { todos: theNewArray }`